### PR TITLE
refactor: Reworked `HitSelector` in Examples

### PR DESCRIPTION
**Pull request by andiwand**
The motivations is that one can select hits to constrain the reconstruction to a subset of them. This can be especially useful for studies which look at different pile-up. Pile-up vertices have consecutive primary vertex IDs so one can simulate for example pile-up 200 and then reconstruct only a subset of them.

Bocked by
- https://github.com/acts-project/acts/pull/3837

## Summary by CodeRabbit

- **New Features**
	- Introduced the `HitSelector` class for enhanced hit selection based on various criteria such as position, time, energy loss, and primary vertex ID.
	- Added new parameters to the `HitSelector` algorithm, allowing for more comprehensive filtering of hits.

- **Bug Fixes**
	- Improved validation logic in the `ParticleSelector` class to include primary vertex ID checks.

- **Documentation**
	- Updated header files to reflect the new structure and functionality of the `HitSelector`.
  
- **Chores**
	- Removed outdated `HitSelector` files from the utilities library to streamline the codebase.

### CODE REVIEW

File: Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/HitSelector.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    const bool validHit = validParticle && validX && validY && validZ &&
+                          validR && validTime && validEnergyLoss &&
+                          validPrimaryVertexId;
+    if (validHit) {
+      unorderedHits.push_back(hit);
+    }
+  }
+
+  SimHitContainer selectedHits(unorderedHits.begin(), unorderedHits.end());
```
Comment (benjaminhuth) [inline Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/HitSelector.cpp:76]: In principle one could optimize the construction by sorting the vector and provide this `boost::container::sorted_range_t` tag (or similar), might be benefitial in for very high pile-up situations with a lot of hits...

Comment (benjaminhuth) [reply Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/HitSelector.cpp:76]: And I think we can reserve the space in the `SimHitContainer`, right?

File: Examples/Framework/include/ActsExamples/EventData/Measurement.hpp [hunk 2]
```diff
@@ -138,6 +129,11 @@ class MeasurementContainer {
     return getMeasurement<Size>(addMeasurement(Size, geometryId));
   }
 
+  template <MeasurementConcept OtherDerived>
```
Comment (benjaminhuth) [inline Examples/Framework/include/ActsExamples/EventData/Measurement.hpp:132]: This seems unrelated?

**Review by benjaminhuth — COMMENTED** [review]
Looks good, just a few comments.

**Review by benjaminhuth — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [1701308e2f852e48fd379d7c887c2efec8fc36df]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575989)||
|:green_circle: | [run_art_test: test_run4_acts_vertex_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575988)||
|:green_circle: | [run_art_test: test_run4_ttbar_PU200](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575987)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575986)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575985)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575984)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575983)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575982)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575981)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575980)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575979)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflowHgtd](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575978)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575977)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575976)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575975)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575974)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575973)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575972)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575971)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575970)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575969)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575968)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575967)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575966)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575965)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575964)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575963)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575962)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575961)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575960)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575959)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575958)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575957)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575956)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575955)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/46575954)||

**Pull request merged**